File: JavaScript/Analysis/Market_Monitor.js
// MarketMonitor.js

import logger from './Logger.js'; //  Logger.js handles logging
import { LinearRegression } from 'ml-regression'; // You may need to install ml-regression
import { TTLCache } from 'ttl-cache'; // You may need to install ttl-cache
import joblib from 'joblib'; // Note: joblib is Python-specific. Use a similar library or implement serialization.
import { Decimal } from 'decimal.js';

class MarketMonitor {
    /**
     * The MarketMonitor class is responsible for monitoring and analyzing market data,
     * updating predictive models, and assessing market conditions such as volatility
     * and liquidity for specified tokens.
     */

    // Class-level constants defining intervals and thresholds
    MODEL_UPDATE_INTERVAL = 3600; // Update model every hour (in seconds)
    VOLATILITY_THRESHOLD = 0.05; // 5% standard deviation as volatility threshold
    LIQUIDITY_THRESHOLD = 100000; // $100,000 in 24h volume as liquidity threshold

    /**
     * Initializes the MarketMonitor instance.
     *
     * @param {Web3} web3 - The Web3 instance for blockchain interactions.
     * @param {Configuration} configuration - The configuration instance.
     * @param {APIConfig} apiconfig - The API configuration instance.
     */
    constructor(web3, configuration, apiconfig) {
        this.web3 = web3;
        this.configuration = configuration;
        this.apiconfig = apiconfig;

        // Initialize the Linear Regression model for price prediction
        this.price_model = new LinearRegression();
        this.model_last_updated = 0; // Timestamp of the last model update

        // Initialize a cache for price data with a TTL of 5 minutes and max size of 1000 entries
        this.price_cache = new TTLCache({ ttl: 300, max: 1000 });

        // Paths for saving/loading the ML model and training data
        this.model_path = this.configuration.ML_MODEL_PATH;
        this.training_data_path = this.configuration.ML_TRAINING_DATA_PATH;

        // Lock to ensure that model updates are thread-safe and prevent concurrent updates
        this.model_lock = false;

        // Asynchronously load existing model and training data if available
        this.load_model();

        logger.debug("MarketMonitor initialized with enhanced configuration.");
    }

    /**
     * Asynchronously load the machine learning model and training data from disk.
     * Utilizes a lock to prevent concurrent access during loading.
     */
    async load_model() {
        if (this.model_lock) return;
        this.model_lock = true;
        try {
            const fs = await import('fs/promises');
            if (await this._file_exists(this.model_path) && await this._file_exists(this.training_data_path)) {
                const data = await joblib.load(this.model_path);
                this.price_model = data.model;
                this.model_last_updated = data.model_last_updated || 0;
                logger.info("ML model and training data loaded successfully.");
            } else {
                logger.info("No existing ML model found. Starting fresh.");
            }
        } catch (error) {
            logger.error(`Failed to load ML model: ${error.message}`);
        } finally {
            this.model_lock = false;
        }
    }

    /**
     * Saves the machine learning model and training data to disk.
     * Utilizes a lock to ensure thread-safe operations during saving.
     */
    async save_model() {
        if (this.model_lock) return;
        this.model_lock = true;
        try {
            const fs = await import('fs/promises');
            const data = {
                model: this.price_model,
                model_last_updated: this.model_last_updated,
            };
            await joblib.dump(data, this.model_path);
            logger.info(`ML model saved to ${this.model_path}.`);
        } catch (error) {
            logger.error(`Failed to save ML model: ${error.message}`);
        } finally {
            this.model_lock = false;
        }
    }

    /**
     * Checks if a file exists.
     *
     * @param {string} path - The file path.
     * @returns {Promise<boolean>} - True if exists, else False.
     */
    async _file_exists(path) {
        const fs = await import('fs/promises');
        try {
            await fs.access(path);
            return true;
        } catch {
            return false;
        }
    }

    /**
     * Periodically train the ML model based on the defined interval.
     * This coroutine runs indefinitely, checking if the model needs an update
     * and triggering the update process accordingly.
     *
     * @param {string} token_symbol - The symbol of the token to update the model for.
     */
    async periodic_model_training(token_symbol) {
        while (true) {
            try {
                const current_time = Date.now() / 1000; // Convert to seconds
                if (current_time - this.model_last_updated > this.MODEL_UPDATE_INTERVAL) {
                    logger.debug(`Time to update ML model for ${token_symbol}.`);
                    await this._update_price_model(token_symbol);
                }
                await this._sleep(60000); // Check every minute
            } catch (error) {
                logger.error(`Error in periodic model training: ${error.message}`);
                await this._sleep(60000); // Wait a minute before retrying
            }
        }
    }

    /**
     * Starts the periodic model training for a specific token.
     *
     * @param {string} token_symbol - The symbol of the token to train the model for.
     */
    async start_periodic_training(token_symbol) {
        this.periodic_model_training(token_symbol);
    }

    /**
     * Updates the price prediction model with new historical data for a specific token.
     * Fetches historical prices, prepares training data, fits the model, and saves it.
     *
     * @param {string} token_symbol - The symbol of the token to update the model for.
     */
    async _update_price_model(token_symbol) {
        if (this.model_lock) return;
        this.model_lock = true;
        try {
            const prices = await this.fetch_historical_prices(token_symbol);
            if (prices.length > 10) {
                // Prepare training data with time steps as features
                const X = prices.map((_, index) => [index]);
                const y = prices;

                // If training data exists, load and append new data
                let X_combined = X;
                let y_combined = y;
                if (await this._file_exists(this.training_data_path)) {
                    const csv = await this.web3.utils.toHex(await this.apiconfig._load_abi(this.training_data_path));
                    // Implement CSV parsing as needed
                    // icity,  existing_data is loaded as arrays
                    const existing_data = []; // 
                    X_combined = existing_data.X.concat(X);
                    y_combined = existing_data.y.concat(y);
                }

                // Fit the Linear Regression model with the new data
                this.price_model = new LinearRegression(X_combined, y_combined);
                this.model_last_updated = Date.now() / 1000;
                logger.info(`ML model updated and trained for ${token_symbol}.`);

                // Persist the updated model to disk
                await this.save_model();
            } else {
                logger.debug(`Not enough price data to train the model for ${token_symbol}.`);
            }
        } catch (error) {
            logger.error(`Error updating price model: ${error.message}`);
        } finally {
            this.model_lock = false;
        }
    }

    /**
     * Assesses various market conditions for a given token, such as volatility,
     * trend direction, and liquidity.
     *
     * @param {string} token_address - The blockchain address of the token to check.
     * @returns {Object} - A dictionary containing the evaluated market conditions.
     */
    async check_market_conditions(token_address) {
        const market_conditions = {
            high_volatility: false,
            bullish_trend: false,
            bearish_trend: false,
            low_liquidity: false,
        };
        const token_symbol = await this.apiconfig.getTokenSymbol(this.web3, token_address);
        if (!token_symbol) {
            logger.debug(`Cannot get token symbol for address ${token_address}!`);
            return market_conditions;
        }

        const prices = await this.fetch_historical_prices(token_symbol, 1);
        if (prices.length < 2) {
            logger.debug(`Not enough price data to analyze market conditions for ${token_symbol}`);
            return market_conditions;
        }

        // Calculate volatility and assess if it exceeds the threshold
        const volatility = this._calculate_volatility(prices);
        if (volatility > this.VOLATILITY_THRESHOLD) {
            market_conditions.high_volatility = true;
        }
        logger.debug(`Calculated volatility for ${token_symbol}: ${volatility}`);

        // Determine trend based on the moving average
        const moving_average = prices.reduce((a, b) => a + b, 0) / prices.length;
        if (prices[prices.length - 1] > moving_average) {
            market_conditions.bullish_trend = true;
        } else if (prices[prices.length - 1] < moving_average) {
            market_conditions.bearish_trend = true;
        }

        // Assess liquidity based on trading volume
        const volume = await this.get_token_volume(token_symbol);
        if (volume < this.LIQUIDITY_THRESHOLD) {
            market_conditions.low_liquidity = true;
        }

        return market_conditions;
    }

    /**
     * Calculates the volatility of a list of prices as the standard deviation of returns.
     *
     * @param {Array<number>} prices - A list of historical prices.
     * @returns {number} - The calculated volatility.
     */
    _calculate_volatility(prices) {
        const price_changes = [];
        for (let i = 1; i < prices.length; i++) {
            price_changes.push((prices[i] - prices[i - 1]) / prices[i - 1]);
        }
        const mean = price_changes.reduce((a, b) => a + b, 0) / price_changes.length;
        const variance = price_changes.reduce((a, b) => a + Math.pow(b - mean, 2), 0) / price_changes.length;
        const std_dev = Math.sqrt(variance);
        return std_dev;
    }

    /**
     * Fetches historical price data for a given token symbol.
     *
     * @param {string} token_symbol - The symbol of the token to fetch prices for.
     * @param {number} [days=30] - Number of days to fetch prices for.
     * @returns {Array<number>} - A list of historical prices.
     */
    async fetch_historical_prices(token_symbol, days = 30) {
        const cache_key = `historical_prices_${token_symbol}_${days}`;
        if (this.price_cache.has(cache_key)) {
            logger.debug(`Returning cached historical prices for ${token_symbol}.`);
            return this.price_cache.get(cache_key);
        }

        const prices = await this._fetch_from_services(
            async (service) => await this.apiconfig.fetchHistoricalPrices(token_symbol, days),
            `historical prices for ${token_symbol}`
        );

        if (prices && prices.length > 0) {
            this.price_cache.set(cache_key, prices);
            return prices;
        }

        return [];
    }

    /**
     * Retrieves the 24-hour trading volume for a specified token.
     *
     * @param {string} token_symbol - The symbol of the token to fetch volume for.
     * @returns {number} - The 24-hour trading volume in USD.
     */
    async get_token_volume(token_symbol) {
        const cache_key = `token_volume_${token_symbol}`;
        if (this.price_cache.has(cache_key)) {
            logger.debug(`Returning cached trading volume for ${token_symbol}.`);
            return this.price_cache.get(cache_key);
        }

        const volume = await this._fetch_from_services(
            async (service) => await this.apiconfig.getTokenVolume(token_symbol),
            `trading volume for ${token_symbol}`
        );

        if (volume !== null) {
            this.price_cache.set(cache_key, volume);
            return volume;
        }

        return 0.0;
    }

    /**
     * Helper method to fetch data from multiple configured services.
     *
     * @param {Function} fetch_func - The function to fetch data from a service.
     * @param {string} description - Description of the data being fetched.
     * @returns {any} - The fetched data or null if all services fail.
     */
    async _fetch_from_services(fetch_func, description) {
        const services = Object.keys(this.apiconfig.apiConfigs);
        for (const service of services) {
            try {
                logger.debug(`Fetching ${description} using ${service}...`);
                const result = await fetch_func(service);
                if (result) {
                    return result;
                }
            } catch (error) {
                logger.warn(`Failed to fetch ${description} using ${service}: ${error.message}`);
            }
        }
        logger.warn(`Failed to fetch ${description}.`);
        return null;
    }

    /**
     * Predicts the next price movement for a given token symbol using the ML model.
     *
     * @param {string} token_symbol - The symbol of the token to predict price movement for.
     * @returns {number} - The predicted price movement.
     */
    async predict_price_movement(token_symbol) {
        const current_time = Date.now() / 1000; // Convert to seconds
        if (current_time - this.model_last_updated > this.MODEL_UPDATE_INTERVAL) {
            logger.debug(`Model needs updating for ${token_symbol}. Triggering update.`);
            await this._update_price_model(token_symbol);
        }

        const prices = await this.fetch_historical_prices(token_symbol, 1);
        if (!prices.length) {
            logger.debug(`No recent prices available for ${token_symbol}.`);
            return 0.0;
        }

        try {
            // Predict the next price based on the current data
            const X_pred = [[prices.length]];
            const predicted_price = this.price_model.predict(X_pred)[0];
            logger.debug(`Price prediction for ${token_symbol}: ${predicted_price}`);
            return predicted_price;
        } catch (error) {
            logger.error(`Error predicting price movement: ${error.message}`);
            return 0.0;
        }
    }

    /**
     * Determines if there's an arbitrage opportunity based on the target transaction.
     *
     * @param {Object} target_tx - The target transaction dictionary.
     * @returns {boolean} - True if an arbitrage opportunity is detected, else False.
     */
    async is_arbitrage_opportunity(target_tx) {
        const decoded_tx = await this.decode_transaction_input(target_tx.input, this.web3.utils.toChecksumAddress(target_tx.to));
        if (!decoded_tx) {
            return false;
        }

        const path = decoded_tx.params.path || [];
        if (path.length < 2) {
            return false;
        }

        const token_address = path[path.length - 1]; // The token being bought
        const token_symbol = await this.apiconfig.getTokenSymbol(this.web3, token_address);
        if (!token_symbol) {
            return false;
        }

        // Fetch real-time prices from different services
        const prices = await this._get_prices_from_services(token_symbol);
        if (prices.length < 2) {
            return false;
        }

        // Calculate the difference and percentage to assess arbitrage potential
        const price_difference = Math.abs(prices[0] - prices[1]);
        const average_price = prices.reduce((a, b) => a + b, 0) / prices.length;
        if (average_price === 0) {
            return false;
        }

        const price_difference_percentage = price_difference / average_price;
        if (price_difference_percentage > 0.01) { // Arbitrage threshold set at 1%
            logger.debug(`Arbitrage opportunity detected for ${token_symbol}`);
            return true;
        }
        return false;
    }

    /**
     * Retrieves real-time prices for a token from different services.
     *
     * @param {string} token_symbol - The symbol of the token to get prices for.
     * @returns {Array<number>} - A list of real-time prices from various services.
     */
    async _get_prices_from_services(token_symbol) {
        const prices = [];
        const services = Object.keys(this.apiconfig.apiConfigs);
        for (const service of services) {
            try {
                const price = await this.apiconfig.getRealTimePrice(token_symbol.toLowerCase());
                if (price !== null) {
                    prices.push(price);
                }
            } catch (error) {
                logger.warn(`Failed to get price from ${service}: ${error.message}`);
            }
        }
        return prices;
    }
}

export default MarketMonitor;

File: python/analysis/market_monitor.py
import time
import numpy as np
from sklearn.linear_model import LinearRegression
from cachetools import TTLCache
from typing import Any, Dict, List, Optional, Union, Tuple
from web3 import AsyncWeb3
import logging
from decimal import Decimal

from configuration.api_config import API_Config
from configuration.configuration import Configuration

logger = logging.getLogger(__name__)

# Add new cache settings
CACHE_SETTINGS = {
    'price': {'ttl': 300, 'size': 1000},
    'volume': {'ttl': 900, 'size': 500},
    'volatility': {'ttl': 600, 'size': 200}
}

class Market_Monitor:
    """Advanced market monitoring system for real-time analysis and prediction."""

    # Class Constants
    MODEL_UPDATE_INTERVAL: int = 3600  # Update model every hour
    VOLATILITY_THRESHOLD: float = 0.05  # 5% standard deviation
    LIQUIDITY_THRESHOLD: int = 100_000  # $100,000 in 24h volume

    def __init__(
        self,
        web3: AsyncWeb3,
        configuration: Optional[Configuration],
        api_config: Optional[API_Config],
        transaction_core: Optional[Any] = None,  # Add this parameter
    ) -> None:
        """Initialize Market Monitor with required components."""
        self.web3 = web3
        self.configuration = configuration
        self.api_config = api_config
        self.transaction_core = transaction_core  # Store transaction_core reference
        self.price_model = LinearRegression()
        self.model_last_updated = 0

        # Add separate caches for different data types
        self.caches = {
            'price': TTLCache(maxsize=CACHE_SETTINGS['price']['size'], 
                            ttl=CACHE_SETTINGS['price']['ttl']),
            'volume': TTLCache(maxsize=CACHE_SETTINGS['volume']['size'], 
                             ttl=CACHE_SETTINGS['volume']['ttl']),
            'volatility': TTLCache(maxsize=CACHE_SETTINGS['volatility']['size'], 
                                 ttl=CACHE_SETTINGS['volatility']['ttl'])
        }

    async def check_market_conditions(
        self, 
        token_address: str
    ) -> Dict[str, bool]:
        """
        Analyze current market conditions for a given token.
        
        Args:
            token_address: Token contract address
            
        Returns:
            Dictionary containing market condition indicators
        """
        market_conditions = {
            "high_volatility": False,
            "bullish_trend": False,
            "bearish_trend": False,
            "low_liquidity": False,
        }
        token_symbol = await self.api_config.get_token_symbol(self.web3, token_address)
        if not token_symbol:
            logger.debug(f"Cannot get token symbol for address {token_address}!")
            return market_conditions

        prices = await self.get_price_data(token_symbol, data_type='historical', timeframe=1)
        if len(prices) < 2:
            logger.debug(f"Not enough price data to analyze market conditions for {token_symbol}")
            return market_conditions

        volatility = self._calculate_volatility(prices)
        if volatility > self.VOLATILITY_THRESHOLD:
            market_conditions["high_volatility"] = True
        logger.debug(f"Calculated volatility for {token_symbol}: {volatility}")

        moving_average = np.mean(prices)
        if prices[-1] > moving_average:
            market_conditions["bullish_trend"] = True
        elif prices[-1] < moving_average:
            market_conditions["bearish_trend"] = True

        volume = await self.get_token_volume(token_symbol)
        if volume < self.LIQUIDITY_THRESHOLD:
            market_conditions["low_liquidity"] = True

        return market_conditions

    # Price Analysis Methods
    async def predict_price_movement(
        self, 
        token_symbol: str
    ) -> float:
        """
        Predict future price movement using linear regression model.
        
        Args:
            token_symbol: Token symbol to analyze
            
        Returns:
            Predicted price value
        """
        current_time = time.time()
        if current_time - self.model_last_updated > self.MODEL_UPDATE_INTERVAL:
            await self._update_price_model(token_symbol)
        prices = await self.get_price_data(token_symbol, data_type='historical', timeframe=1)
        if not prices:
            logger.debug(f"No recent prices available for {token_symbol}.")
            return 0.0
        next_time = np.array([[len(prices)]])
        predicted_price = self.price_model.predict(next_time)[0]
        logger.debug(f"Price prediction for {token_symbol}: {predicted_price}")
        return float(predicted_price)

    # Market Data Methods
    async def get_price_data(self, *args, **kwargs):
        """Use centralized price fetching from API_Config."""
        return await self.api_config.get_token_price_data(*args, **kwargs)

    async def get_token_volume(self, token_symbol: str) -> float:
        """
        Get the 24-hour trading volume for a given token symbol.

        :param token_symbol: Token symbol to fetch volume for
        :return: 24-hour trading volume
        """
        cache_key = f"token_volume_{token_symbol}"
        if cache_key in self.caches['volume']:
            logger.debug(f"Returning cached trading volume for {token_symbol}.")
            return self.caches['volume'][cache_key]

        volume = await self._fetch_from_services(
            lambda _: self.api_config.get_token_volume(token_symbol),
            f"trading volume for {token_symbol}"
        )
        if volume is not None:
            self.caches['volume'][cache_key] = volume
        return volume or 0.0

    async def _fetch_from_services(self, fetch_func, description: str) -> Optional[Union[List[float], float]]:
        """
        Helper method to fetch data from multiple services.

        :param fetch_func: Function to fetch data from a service
        :param description: Description of the data being fetched
        :return: Fetched data or None
        """
        for service in self.api_config.api_configs.keys():
            try:
                logger.debug(f"Fetching {description} using {service}...")
                result = await fetch_func(service)
                if result:
                    return result
            except Exception as e:
                logger.warning(f"failed to fetch {description} using {service}: {e}")
        logger.warning(f"failed to fetch {description}.")
        return None

    # Helper Methods
    def _calculate_volatility(
        self, 
        prices: List[float]
    ) -> float:
        """
        Calculate price volatility using standard deviation of returns.
        
        Args:
            prices: List of historical prices
            
        Returns:
            Volatility measure as float
        """
        prices_array = np.array(prices)
        returns = np.diff(prices_array) / prices_array[:-1]
        return np.std(returns)

    async def _update_price_model(self, token_symbol: str) -> None:
        """
        Update the price prediction model.

        :param token_symbol: Token symbol to update the model for
        """
        prices = await self.get_price_data(token_symbol, data_type='historical')
        if len(prices) > 10:
            X = np.arange(len(prices)).reshape(-1, 1)
            y = np.array(prices)
            self.price_model.fit(X, y)
            self.model_last_updated = time.time()

    async def is_arbitrage_opportunity(self, target_tx: Dict[str, Any]) -> bool:
        """Use transaction_core for decoding but avoid circular dependencies."""
        if not self.transaction_core:
            logger.warning("Transaction core not initialized, cannot check arbitrage")
            return False

        try:
            decoded_tx = await self.transaction_core.decode_transaction_input(
                target_tx["input"], target_tx["to"]
            )
            # ...existing code...
        except Exception as e:
            logger.error(f"Error checking arbitrage opportunity: {e}")
            return False

    async def _get_prices_from_services(self, token_symbol: str) -> List[float]:
        """
        Get real-time prices from different services.

        :param token_symbol: Token symbol to get prices for
        :return: List of real-time prices
        """
        prices = []
        for service in self.api_config.api_configs.keys():
            try:
                price = await self.api_config.get_real_time_price(token_symbol)
                if price is not None:
                    prices.append(price)
            except Exception as e:
                logger.warning(f"failed to get price from {service}: {e}")
        return prices

    async def decode_transaction_input(self, *args, **kwargs):
        """Use centralized transaction decoding from Transaction_Core."""
        return await self.transaction_core.decode_transaction_input(*args, **kwargs)

    async def initialize(self) -> None:
        """Initialize market monitor components."""
        try:
            self.price_model = LinearRegression()
            self.model_last_updated = 0
            logger.debug("Market Monitor initialized ✅")
        except Exception as e:
            logger.critical(f"Market Monitor initialization failed: {e}")
            raise

    async def stop(self) -> None:
        """Clean up resources and stop monitoring."""
        try:
            # Clear caches and clean up resources
            for cache in self.caches.values():
                cache.clear()
            logger.debug("Market Monitor stopped.")
        except Exception as e:
            logger.error(f"Error stopping Market Monitor: {e}")
